src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java

```diff
@@ -0,0 +1,408 @@
+     */
+    @Override
+    public Iterator<VariantContext> query( final SimpleInterval interval ) {
+        return getMergedIteratorFromDataSources(ds -> ds.queryAndPrefetch(interval).iterator());
+    }
+
+    /**
+     * Close any existing iterator, create a new iterator and update the local cached iterator reference.
+     * @param iteratorFromSource function to retrieve individual iterator, to be applied to each data source
+     * @return
+     */
+    private Iterator<VariantContext> getMergedIteratorFromDataSources(
+            final Function<FeatureDataSource<VariantContext>, Iterator<VariantContext>> iteratorFromSource) {
+        validateFeatureSources();
+
+        // Tribble documentation states that having multiple iterators open simultaneously over the same FeatureReader
+        // results in undefined behavior
+        closeOpenIterationIfNecessary();
+
+        List<CloseableIterator<VariantContext>> iterators = new ArrayList<>(1);
+        featureDataSources.forEach(ds -> iterators.add(getCloseableIteratorWrapper(iteratorFromSource.apply((ds)))));
+
+        VariantContextComparator varComparator = new VariantContextComparator(getSequenceDictionary());
+        currentIterator = new MergingIterator<>(varComparator, iterators);
+        return currentIterator;
+    }
+
+    /**
+     * Get the logical name of this data source.
+     *
+     * @return the logical name of this data source
+     */
+    public String getName() {
+        validateFeatureSources();
+        return "MultiVariantDataSource: ("
+                + Utils.join(", ", featureDataSources.stream().map(fds -> fds.getName()).collect(Collectors.toList()))
+                + ")";
+    }
+
+    /**
+     * Gets the header associated with this data source
+     *
+     * @return header associated with this data source as an Object
+     */
+    public VCFHeader getHeader() {
+        validateFeatureSources();
+        if (mergedHeader == null) { // merge and cache the resulting header
+            List<VCFHeader> headers = featureDataSources
+                    .stream()
+                    .map(ds -> (VCFHeader) ds.getHeader())
+                    .filter(h -> h != null)
+                    .collect(Collectors.toList());
+
+            // Now merge the headers using htsjdk, which is pretty promiscuous, and which only works properly
+            // because of the cross-dictionary validation done in validateAllSequenceDictionaries.
+            mergedHeader = headers.size() > 1 ?
+                    new VCFHeader(VCFUtils.smartMergeHeaders(headers, true)) :
+                    headers.get(0);
+        }
+        return mergedHeader;
+    }
+
+    /**
+     * Permanently close this data source, invalidating any open iteration over it, and making it invalid for future
+     * iterations and queries.
+     */
+    @Override
+    public void close() {
+        validateFeatureSources();
+        featureDataSources.forEach(dataSource -> dataSource.close());
```

Comment:
Don't you need to close out any merging iterators that were handed out to clients?